src/v/iceberg/metadata_query.cc

```diff
@@ -0,0 +1,236 @@
+          return p_spec.spec_id == id;
+      });
+
+    if (p_spec_it == table.partition_specs.end()) {
+        return iceberg::query_executor::errc::table_metadata_inconsistency;
+    }
+
+    return partition_key_type::create(*p_spec_it, *s_it);
+}
+
+} // namespace
+
+template<result_type ResultT>
+ss::future<checked<
+  typename internal::results_container<ResultT>::type,
+  query_executor::errc>>
+query_executor::do_execute_query(const metadata_query<ResultT>& query) const {
+    results_collector<ResultT> collector;
+
+    if (!table_.snapshots) {
+        co_return std::move(collector).release();
+    }
+    const auto& snapshots = table_.snapshots.value();
+    // simple DFS scan over metadata
+    for (const auto& s : snapshots) {
+        if (!matches(query, s)) {
+            continue;
+        }
+        // snapshot matches, check if manifest list is needed, otherwise simply
+        // collect the snapshot
+        if (needs_manifest_list(query)) {
+            auto m_list_result = co_await io_.download_manifest_list(
+              manifest_list_path{s.manifest_list_path});
+
+            if (m_list_result.has_error()) {
+                vlog(
+                  log.warn,
+                  "error downloading manifest list from {} - {}",
+                  s.manifest_list_path,
+                  fmt::underlying_t<metadata_io::errc>(m_list_result.error()));
+                co_return errc::metadata_io_error;
+            }
+
+            for (auto& manifest_file : m_list_result.value().files) {
+                if (!matches(query, manifest_file)) {
+                    continue;
+                }
+
+                if (needs_manifest(query)) {
+                    auto pk_result = make_partition_key_type(
+                      table_, s, manifest_file);
+                    if (pk_result.has_error()) {
+                        co_return pk_result.error();
+                    }
+                    auto m_result = co_await io_.download_manifest(
+                      manifest_path(manifest_file.manifest_path),
+                      std::move(pk_result.value()));
+
+                    if (m_result.has_error()) {
+                        vlog(
+                          log.warn,
+                          "error downloading manifest from {} - {}",
+                          manifest_file.manifest_path,
+                          fmt::underlying_t<metadata_io::errc>(
+                            m_result.error()));
+                        co_return errc::metadata_io_error;
+                    }
+                    if (matches(query, m_result.value())) {
+                        if (collector.collect(
+                              s, manifest_file, m_result.value())) {
+                            break;
+                        }
+                    }
+                } else {
+                    if (collector.collect(s, manifest_file)) {
+                        break;
+                    }
+                }
```

Comment:
Having the else clause here so far away makes things harder to read for me (especially because the nesting is quite deep). Consider something like:
```
if (!matches(...)) {
  continue;
}
if (!needs_manifest(...)) {
  if (collector.collect(...)) {
    break;
  }
  continue;
}
// rest of logic unnested.
```